### PR TITLE
feat(test): format user code output

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno/issues/14192

This commit changes "deno test" to better denote user output coming
from test cases. 

This is done by printing "---- output ----" and "---- output end ----" 
markers if an output is produced. The output from "console" and 
"Deno.core.print" is captured, as well as direct writes to "Deno.stdout"
and "Deno.stderr".

To achieve that new APIs were added to "deno_core" crate, that allow
to replace an existing resource with a different one (while keeping resource
ids intact). Resources for stdout and stderr are replaced by pipes.

Screenshot 2022-04-14 at 19 20 44

### CODE REVIEW

**Comment by bartlomieju** [comment]
Before landing, the same should be done for benching code.

File: cli/tools/test.rs [hunk 1]
```diff
@@ -314,9 +319,24 @@ impl TestReporter for PrettyTestReporter {
...
+    }
+
+    if !self.did_have_user_output {
+      self.did_have_user_output = true;
+      println!();
+      println!("{}", colors::gray("-------output:-------"));
+    }
+    match output {
+      TestOutput::PrintStdout(line) | TestOutput::PrintStderr(line) => {
+        print!("{}", line)
```
Comment (bartlomieju) [inline cli/tools/test.rs:333]: This should probably be `print!` for `PrintStdout` and `eprint!` for `PrintStderr`

File: cli/tools/test.rs [hunk 2]
```diff
@@ -81,7 +81,10 @@ pub struct TestDescription {
 #[serde(rename_all = "camelCase")]
 pub enum TestOutput {
   // TODO(caspervonb): add stdout and stderr redirection.
```
Comment (bartlomieju) [inline cli/tools/test.rs:83]: Remove

File: cli/lsp/testing/execution.rs [hunk 3]
```diff
@@ -753,14 +753,18 @@ impl test::TestReporter for LspTestReporter {
...
+      | test::TestOutput::PrintStderr(value) => {
         self.progress(lsp_custom::TestRunProgressMessage::Output {
           value: value.replace('\n', "\r\n"),
           test,
           // TODO(@kitsonk) test output should include a location
           location: None,
         })
       }
+      test::TestOutput::Stdout(_) | test::TestOutput::Stderr(_) => {
+        // todo(@dsherret): handle
```
Comment (bartlomieju) [inline cli/lsp/testing/execution.rs:766]: Fix before landing

**Comment by bartlomieju** [comment]
> Before landing, the same should be done for benching code.

I will actually address this in a follow up PR.

**Review by dsherret — APPROVED** [review]
LGTM

**Pull request merged**